File: task4/task.go
package main

import (
	"fmt"
)

func main() {

	var n, m int

	fmt.Scan(&n, &m)

	a := make([]int, n)

	for i := 0; i < n; i++ {
		fmt.Scan(&a[i])
	}

	pavel := 0
	vika := 0

	pos := 0

	// Предыдущие ходы игроков
	prevPavel := -1
	prevVika := -1

	pavelTurn := true

	for pos < n {

		bestSum := -1000000000
		bestCount := 1

		currentSum := 0

		for k := 1; k <= m && pos+k <= n; k++ {

			// Проверка ограничения
			if pavelTurn && k == prevPavel {
				continue
			}

			if !pavelTurn && k == prevVika {
				continue
			}

			currentSum += a[pos+k-1]

			// Выбор максимальной суммы
			if currentSum > bestSum {
				bestSum = currentSum
				bestCount = k
			}
		}

		currentSum = 0

		for i := 0; i < bestCount; i++ {
			currentSum += a[pos]
			pos++
		}

		if pavelTurn {
			pavel += currentSum
			prevPavel = bestCount
		} else {
			vika += currentSum
			prevVika = bestCount
		}

		pavelTurn = !pavelTurn
	}

	if pavel > vika {
		fmt.Println(1)
	} else {
		fmt.Println(0)
	}
	
	fmt.Println()
}

File: task4/task.cpp
#include <iostream>
#include <vector>

using namespace std;

int main()
{
    int n, m;
    cin >> n >> m;

    vector<int> a(n);

    for (int i = 0; i < n; i++)
        cin >> a[i];

    int pavel = 0;
    int vika = 0;

    int pos = 0;

    // Предыдущие ходы игроков
    int prevPavel = -1;
    int prevVika = -1;

    bool pavelTurn = true;

    while (pos < n)
    {
        int bestSum = -1000000000;
        int bestCount = 1;

        int currentSum = 0;

        for (int k = 1; k <= m && pos + k <= n; k++)
        {
            // Проверка ограничения
            if (pavelTurn && k == prevPavel)
                continue;

            if (!pavelTurn && k == prevVika)
                continue;

            currentSum += a[pos + k - 1];

            // Выбираем максимальную сумму
            // При равенстве — минимальное количество
            if (currentSum > bestSum)
            {
                bestSum = currentSum;
                bestCount = k;
            }
        }

        currentSum = 0;

        for (int i = 0; i < bestCount; i++)
        {
            currentSum += a[pos];
            pos++;
        }

        if (pavelTurn)
        {
            pavel += currentSum;
            prevPavel = bestCount;
        }
        else
        {
            vika += currentSum;
            prevVika = bestCount;
        }

        pavelTurn = !pavelTurn;
    }

    if (pavel > vika)
        cout << 1 << endl;
    else
        cout << 0 << endl;

      cout << endl;

    return 0;
}
